### PR TITLE
Token fetcher re-implemented

### PR DESCRIPTION
**Pull request by semyonoskin**

### CODE REVIEW

**Review by GusevTimofey — CHANGES_REQUESTED** [review]

File: modules/markets-api/src/main/scala/org/ergoplatform/dex/markets/services/TokenFetcher.scala [hunk 1]
```diff
@@ -24,27 +29,47 @@ trait TokenFetcher[F[_]] {
...
+        currTs             <- millis
+        tokens <- if (FiniteDuration(currTs - ts, TimeUnit.MILLISECONDS) > conf.rate)
+                    requestedTokens.flatMap(tokens => tokenRef.set((currTs, tokens)).map(_ => tokens))
+                  else cachedTokens.pure
+      } yield tokens
+
+    private val requestedTokens: F[List[TokenId]] =
+      basicRequest
+        .get(uri"${conf.filePath}")
+        .response(asString)
```
[the remark this answers is not in the document]
Comment (GusevTimofey) [reply modules/markets-api/src/main/scala/org/ergoplatform/dex/markets/services/TokenFetcher.scala:59]: Json file is:
https://github.com/ergolabs/default-token-list/blob/master/default-token-list.json

File: modules/markets-api/src/main/scala/org/ergoplatform/dex/markets/configs/TokenFetcherConfig.scala [hunk 2]
```diff
@@ -4,8 +4,9 @@ import derevo.derive
 import derevo.pureconfig.pureconfigReader
 import tofu.Context
 import tofu.logging.derivation.loggable
+import scala.concurrent.duration.FiniteDuration
 
 @derive(pureconfigReader, loggable)
-case class TokenFetcherConfig(filePath: String)
+case class TokenFetcherConfig(filePath: String, rate: FiniteDuration)
```
Comment (GusevTimofey) [inline modules/markets-api/src/main/scala/org/ergoplatform/dex/markets/configs/TokenFetcherConfig.scala:10]: It's now **url** not **filePath**

File: modules/markets-api/src/main/scala/org/ergoplatform/dex/markets/services/TokenFetcher.scala [hunk 3]
```diff
@@ -24,27 +27,40 @@ trait TokenFetcher[F[_]] {
...
+                  else cachedTokens.pure
+      } yield tokens
+
+    private val requestedTokens: F[List[TokenId]] =
+      basicRequest
+        .get(uri"${conf.url}")
+        .response(asJson[TokenResponse])
+        .send(backend)
+        .absorbError
+        .map(_.tokens.filter(_.network == "ergo").map(tkn => TokenId.fromStringUnsafe(tkn.address)))
```
Comment (GusevTimofey) [inline modules/markets-api/src/main/scala/org/ergoplatform/dex/markets/services/TokenFetcher.scala:60]: It is better to put "ergo" into constants

**Review by GusevTimofey — CHANGES_REQUESTED** [review]

**Comment by GusevTimofey** [comment]
Did you check it if it works or not?

**Comment by semyonoskin** [comment]
Ye, works fine

**Review by GusevTimofey — APPROVED** [review]

File: modules/markets-api/src/main/scala/org/ergoplatform/dex/markets/services/TokenFetcher.scala [hunk 4]
```diff
@@ -24,27 +27,42 @@ trait TokenFetcher[F[_]] {
...
+    private val requestedTokens: F[List[TokenId]] =
+      basicRequest
+        .get(uri"${conf.url}")
+        .response(asJson[TokenResponse])
+        .send(backend)
+        .absorbError
+        .map(_.tokens.filter(_.network == network).map(tkn => TokenId.fromStringUnsafe(tkn.address)))
+
+    final case class Token(network: String, address: String, decimals: Int, name: String, ticker: String)
+    final case class TokenResponse(tokens: List[Token])
```
Comment (oskin1) [inline modules/markets-api/src/main/scala/org/ergoplatform/dex/markets/services/TokenFetcher.scala:65]: Please move case class declarations outside class. E.g to TokenFetcher object level.

**Review by oskin1 — APPROVED** [review]

**Pull request merged**